clock-bound-c/src/lib.rs

```diff
@@ -339,6 +359,71 @@ impl ClockBoundClient {
             after,
         })
     }
+
+    ///Execute `f` and return bounds on execution time
+    pub fn timing<A, F>(&self, f: F) -> Result<(TimingResult, A), (ClockBoundCError, Result<A,F>)>
+        where F: FnOnce() -> A {
+
+        // Get the first timestamps
+
+        // Header
+        // 1st - Version
+        // 2nd - Command Type
+        // 3rd, 4th - Reserved
+        let mut request: [u8; 4] = [1, 1, 0, 0];
+
+        match self.socket.send(&mut request) {
+            Err(e) => return Err((ClockBoundCError::SendMessageError(e), Err(f))),
+            _ => {}
+        }
+        let mut response: [u8; 20] = [0; 20];
+        match self.socket.recv(&mut response) {
+            Err(e) => return Err((ClockBoundCError::ReceiveMessageError(e), Err(f))),
+            _ => {}
+        }
+        let response_earliest_start = NetworkEndian::read_u64(&response[4..12]);
+        let response_earliest_finish = NetworkEndian::read_u64(&response[12..20]);
+
+        // Execute the provided function, f
+        let callback = f();
+
+        // Get the second timestamps
+        let mut request: [u8; 4] = [1, 1, 0, 0];
+        match self.socket.send(&mut request) {
+            Err(e) => return Err((ClockBoundCError::SendMessageError(e), Ok(callback))),
+            _ => {}
+        }
+        let mut response: [u8; 20] = [0; 20];
+        match self.socket.recv(&mut response) {
+            Err(e) => return Err((ClockBoundCError::ReceiveMessageError(e), Ok(callback))),
+            _ => {}
+        }
+        let response_latest_start = NetworkEndian::read_u64(&response[4..12]);
+        let response_latest_finish = NetworkEndian::read_u64(&response[12..20]);
+
+        // Calculate elapsed time between early and latest timestamps
+        let earliest_time_elapsed_midpoint = (response_earliest_finish - response_earliest_start)/2;
+        let latest_time_elapsed_midpoint = (response_latest_finish - response_latest_start)/2;
+
+        let earliest_start = UNIX_EPOCH + Duration::from_nanos(response_earliest_start);
+        let latest_start = UNIX_EPOCH + Duration::from_nanos(response_latest_start);
+        let latest_finish = UNIX_EPOCH + Duration::from_nanos(response_latest_finish);
+
+        // Add half of time elapsed to the start times to get the mid points
+        let duration_start = earliest_start.checked_add(Duration::from_nanos(earliest_time_elapsed_midpoint)).unwrap();
+        let duration_finish = latest_start.checked_add(Duration::from_nanos(latest_time_elapsed_midpoint)).unwrap();
+
+        // Calculates duration between the two midpoints
+        let duration = duration_finish.duration_since(duration_start).unwrap();
+        let ppm = Duration::from_micros(DEFAULT_CLOCK_FREQUENCY);
+
+        Ok((TimingResult{
+            earliest_start,
+            latest_finish,
+            min_execution_time: duration.saturating_sub(ppm),
```

Comment:
This math isn't correct. If adjust by ±1ppm, you want to be multiplying by 0.999999 and 1.000001. Instead you're in effect adding or subtracting one microsecond, regardless of estimated elapsed time.

Reply:
Fixed. Heres a sample output 
```
Earliest start time for the timing request: "2022-02-23 17:47:43.459195163"
Latest finish time for the timing request: "2022-02-23 17:47:43.480670837"
Minimum execution duration of timing request: 89.999µs
Maximum execution duration of timing request: 90µs
```